Login / register fail modal in AuthProvider

SimonBartonPSU opened the pull request:
EDIT: This now is working. It's a somewhat jenky solution, but.... it works! You can try changing the inner `setTimeout` value on line 163 of AuthProvider.js to something really long like 9000 and verify we are able to control time now.

I spent many hours trying to fuss with state management. Could never get better than the 'ever-other' result that Erik was able to achieve. (Using a context within another screen is difficult if it gets more complicated than referencing a single value/function!!!)

Now, we have the two-layered `setTimeout`. The outer one (with 800 as the ms timeout value) basically let's the event queue clear so the LoginScreen or RegisterScreen won't re-render over the modal. Thus, the `showModal` and subsequent inner `setTimeout` which calls `hideModal` is guaranteed to be the last thing to execute, uninterrupted on either Login or Register screen.

Feel free to test and if you find `setTimeout` values that feel better, push a commit! These felt pretty good to me, but I was mostly fed up with this issue once is worked at all.

SimonBartonPSU committed:
Show failure modal for failed login and register

Not quite there. Cannot get the modal to stay for long enough,
particularly on login screen.

seansisson reviewed (commented):
Due to some issues on my Mac I can't review the iOS changes at this time, but I will test the functionality on Android & web app

EHaake committed:
Add authFailed var to Context and listen on it in Login Screen

Currently the modal popup displays every other time a login attempt
fails, but does display for the whole time

EHaake commented:
So I added an `authFailed` state variable that is set to true if the performAuthRequest method doesn't succeed and a method to reset it's value. Then in the Login screen, I'm listening to that value and if it is true, I try to display a `LOGIN_FAILED` modal. However it works every other time a login attempt fails (starting on failed attempt 2). I don't know why this is happening, though I suspect it may be due to the variable not being updated in time.

SimonBartonPSU committed:
Double layered setTimeout to show failure modals

SimonBartonPSU committed:
Remove unused imports

EHaake committed:
Tweak outer timeout to 600ms from 800ms

EHaake reviewed (approved):
Great work! I was going to try a completely different solution where I moved the auth request logic out of the AuthProvider into the Login Component itself which would have been effectively giving up. Thanks for figuring it out this way. Works great. I just tweaked the timing of the outer timeout down 200ms to 600 from 800 which works on all my tested devices: (Android Emulator, Pixel 3a, iPhone 6s Plus). But probably want to give it a try to make sure

SimonBartonPSU commented:
Yeah the 600 timeout seems a bit smoother. Groovy roovy :+1:

KipTheTurtlebear reviewed (approved):
Works for me, the 600ms feels right
